### DISCUSSION TITLE
Serverless Function Execution limit 75% warning when moved from SSR to SSG

### DISCUSSION THREAD

**Discussion by troymcginnis** [vercel/community#596] [Help] [answered]
Hey Vercel, 

### The problem

We’re currently trying to troubleshoot an issue where our Serverless Function Execution times have skyrocketed since moving one of our web apps over from SSR to SSG. We went from about 77 GB-hr to 1,000 GB-hr month over month. Same amount of traffic, same everything, just SSR -> SSG.

<img width="1036" alt="Screen Shot 2022-05-09 at 2 43 27 PM" src="https://user-images.githubusercontent.com/4299551/167503621-3bae4014-d7ec-4c3d-ab9d-870cd774214d.png">

In talking with Vercel support, we’ve gotten a crumb of insight to suggest that the bulk of the execution times are coming from our catch-all route with a previous months usage of 140 GB-hr alone. With some 70 GB-hr being dedicated to a specific cached version of the .json dataset for that route.

![unnamed](https://user-images.githubusercontent.com/4299551/167503300-f2e526a6-1206-4880-9a25-cfe52648c660.png)

### What I’m trying to get a better handle on is:

How do I reduce our GB-hr usage? Is this strictly data transmitted? So, should I just be looking reducing our payload?
We also introduced middleware code here. I’m wondering if that is also something that’s at play here.

How do we get further insights into what caused this spike?

Any and all help would be greatly appreciated.

**Comment by samcx** [comment]
@troymcginnis I recommend reaching out the case again for this. Can you share with us the case #?

Dynamic pages built at request time will incur SFE to build the page. ISR is also a Serverless Function rebuilding your page in the background.

**Reply by troymcginnis** [reply]
@samsisle The case # is 00078370. I reached out on Monday (3 days ago) and haven't heard back yet.

As of today, we've officially crossed over 100% usage and are burning roughly 40–60 GB-hrs daily. So, at $40/100 GB-hrs, we will likely see a bill of over $400 for this. 

Any support would be massively appreciated here.

**Dynamic Pages/ISR:**
My assumption is that we will see a large decrease in SFE by switching from ISR to On-demand Revalidation. Would you reckon this would be a good starting place to trim the fat? Any further suggestions for reducing our SFE?

**Reply by leerob** [reply]
Would highly recommend on-demand ISR, yes. Could greatly reduce usage.

**Reply by troymcginnis** [reply]
Thanks @leerob. I'll be looking to implement on-demand ISR in the coming weeks to this project. Might be a good case study on how powerful on-demand ISR really is 🤞🏼

In the meantime, I continue to welcome any suggestions on further reducing our SFE. Where my mind is currently going is reducing overall page payload and data API response times, etc.

**Reply by troymcginnis** [reply]
@leerob Is there a way from within `getStaticProps` to determine if a request is coming from on-demand ISR or not? I know that from within middleware there's a header of `x-prerender-revalidate` that is being passed in, but I haven't found a way to determine from within `getStaticProps` if the request is from on-demand ISR or not.

**Reply by leerob** [reply]
Could you share more about what you'd like to do different if it's on-demand vs. normal?

**Reply by troymcginnis** [reply]
I’d like to turn off the use of Sanity’s CDN for on-demand vs. a build or standard revalidation. Content is freshest off of the CDN but there are various reasons to stay on the CDN otherwise.

**Reply by troymcginnis** [reply]
Thoughts @leerob ?

**Reply by leerob** [reply]
I'm not sure this will be possible, since `getStaticProps` does not have access to the incoming request.

**Comment by troymcginnis** [comment, marked as answer]
After implementing on-demand ISR successfully on our client project, I wanted to fire over a quick update. The on-demand ISR implementation was launched to production on **June 25, 2022** and we had some usage spikes just prior to launch, so some of the data might be a bit skewed.

## Vercel
We've successfully reduced our Serverless Function Execution (SFE) times dramatically with the implementation of on-demand ISR:
- Current billing period: **582 GB-Hrs**
- June–July period: **1,495 GB-Hrs**
- May–June period: **2,274 GB-Hrs**
- April–May period: **1,670 GB-Hrs**

## Sanity
We have seen an increase in Sanity API usage, but a reduction in CDN API requests and overall bandwidth:
- Current billing period: 
  - CDN API: **26.6k**
  - API: **656.6k**
  - Bandwidth: **75.3 GB**
- June–July period: 
  - CDN API: **1.2m**
  - API: **34.8k**
  - Bandwidth: **118 GB**
- May–June period: 
  - CDN API: **3.7m**
  - API: **904.4k**
  - Bandwidth: **524.6 GB**
- May–June period: 
  - CDN API: **4.2m**
  - API: **237.1k**
  - Bandwidth: **430.6 GB**

Great work on Next.js Vercel team.

-Troy
